qiskit/circuit/library/evolved_operator_ansatz.py

```diff
@@ -240,11 +240,12 @@ def _build(self):
             evolution.compose(self.initial_state, front=True, inplace=True)
 
         # cast global phase to float if it has no free parameters
-        try:
-            evolution.global_phase = float(evolution.global_phase._symbol_expr)
-        except RuntimeError:
-            # expression contains free parameters
-            pass
+        if isinstance(evolution.global_phase, ParameterExpression):
+            try:
+                evolution.global_phase = float(evolution.global_phase._symbol_expr)
+            except RuntimeError:
```

Comment:
Is it definitely only `RuntimeError` we want to catch?  I'm not very familiar with `sympy`, but
```python
>>> import sympy
>>> float(sympy.sympify('x'))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/jake/.miniconda3/envs/qiskit/lib/python3.9/site-packages/sympy/core/expr.py", line 358, in __float__
    raise TypeError("can't convert expression to float")
TypeError: can't convert expression to float
```
and `TypeError` doesn't derive from `RuntimeError`.

Reply:
There is actually more than just sympy in play here. It'll default to symengine everywhere except for windows and 32bit platforms which will use sympy. I would not be surprised if there are different exceptions raised between the two.

Kind of unrelated why `float(evolution.global_phase._symbol_expr)` instead of `float(evolution.global_phase)`?

Reply:
> Kind of unrelated why float(evolution.global_phase._symbol_expr) instead of float(evolution.global_phase)?

Because that might have unbound parameters even though they add up to something non-parameterized:

```python
>>> from qiskit.circuit import Parameter
>>> x = Parameter('x')
>>> expr = x - x
>>> float(expr)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/jul/Qiskit/qiskit-terra/qiskit/circuit/parameterexpression.py", line 421, in __float__
    "cannot be cast to a float.".format(self.parameters)
TypeError: ParameterExpression with unbound parameters ({Parameter(x)}) cannot be cast to a float.
```

> Is it definitely only RuntimeError we want to catch? I'm not very familiar with sympy, but

I was running with symengine which raised a `RuntimeError`. I'll make it catch both.

Reply:
I guess this more raises the question as to whether we should replace `ParameterExpression.__float__` to handle cases where the underlying symbol expression should return a float in these cases or not.  If the behaviour in this PR is anything to go by, that's probably useful, but we can do it in a different PR.